mne/preprocessing/xdawn.py

```diff
@@ -244,60 +356,17 @@ def fit(self, epochs, y=None):
         if self.correct_overlap == 'auto':
             self.correct_overlap = _check_overlapp(epochs)
 
-        # Extract signal covariance
-        if self.signal_cov is None:
-            if self.correct_overlap:
-                sig_data = _construct_signal_from_epochs(epochs)
-            else:
-                sig_data = np.hstack(epochs.get_data())
-            self.signal_cov_ = _regularized_covariance(sig_data, self.reg)
-        elif isinstance(self.signal_cov, Covariance):
-            self.signal_cov_ = self.signal_cov.data
-        elif isinstance(self.signal_cov, np.ndarray):
-            self.signal_cov_ = self.signal_cov
-        else:
-            raise ValueError('signal_cov must be None, a covariance instance '
-                             'or a ndarray')
-
-        # estimates evoked covariance
-        self.evokeds_cov_ = dict()
-        if self.correct_overlap:
-            if epochs.baseline is not None:
-                raise ValueError('Baseline correction must be None if overlap '
-                                 'correction activated')
-            evokeds, toeplitz = least_square_evoked(epochs,
-                                                    return_toeplitz=True)
-        else:
-            evokeds = dict()
-            toeplitz = dict()
-            for eid in epochs.event_id:
-                evokeds[eid] = epochs[eid].average()
-                toeplitz[eid] = 1.0
-        self.evokeds_ = evokeds
-
-        for eid in epochs.event_id:
-            data = np.dot(evokeds[eid].data, toeplitz[eid])
-            self.evokeds_cov_[eid] = _regularized_covariance(data, self.reg)
-
-        # estimates spatial filters
-        for eid in epochs.event_id:
-
-            if self.signal_cov_.shape != self.evokeds_cov_[eid].shape:
-                raise ValueError('Size of signal cov must be the same as the'
-                                 ' number of channels in epochs')
-
-            evals, evecs = linalg.eigh(self.evokeds_cov_[eid],
-                                       self.signal_cov_)
-            evecs = evecs[:, np.argsort(evals)[::-1]]  # sort eigenvectors
-            evecs /= np.sqrt(np.sum(evecs ** 2, axis=0))
-
-            self.filters_[eid] = evecs
-            self.patterns_[eid] = linalg.inv(evecs.T)
-
+        events, epochs_data, event_id = self._preproc_epochs(epochs)
+        self._get_signal_cov(epochs_data)
+        self._fit_xdawn(epochs_data, events[:, -1], event_id, events,
+                        epochs.tmin, epochs.tmax, epochs.info,
+                        epochs.baseline)
         # store some values
         self.ch_names = epochs.ch_names
         self.exclude = list(range(self.n_components, len(self.ch_names)))
```

Comment:
ok, we're trying to keep the compatibility with 2.7, you can just do

`np.arange(self.n_components, len(self.ch_names))`